src/som/interpreter/actors/SPromise.java

```diff
@@ -328,8 +329,74 @@ public static void setSOMClass(final SClass cls) {
       resolverClass = cls;
     }
 
-    public final void onError() {
-      throw new NotYetImplementedException(); // TODO: implement
+    public static void onError(final SAbstractObject exception, final Object wrapped, final SPromise p, final Actor current) {
+      //System.out.println("called on error");
+      // for trace  TODO fix tracing
+      if (VmSettings.PROMISE_RESOLUTION) {
+        ActorExecutionTrace.promiseRuin(p.getPromiseId(), exception);
+      }
+      synchronized (wrapped) {
+        //System.out.println(p.chainedPromise);
+        //System.out.println(p.chainedPromiseExt);
+       handleOrPasstoChain(exception, wrapped, p, current);
+      }
+    }
+
+    public static void handleOrPasstoChain(final SAbstractObject exception, final Object wrapped, final SPromise p, final Actor current){
+      //System.out.println("called handler or pass to chain");
+      if (!runHandlersUnsync(exception, wrapped, p, current)){
+          //if I did not handle the exception ask chain to handle
+          ruinChainedPromisesUnsynch(exception, p, current);
+      }
+    }
+
+    private static boolean runHandlersUnsync(final SAbstractObject exception, final Object wrapped, final SPromise p, final Actor current) {
+        //sSystem.out.println("called run handlers");
+        synchronized (p) {
+          assert p.assertNotCompleted();
+          p.value = exception;
+          p.resolutionState = Resolution.ERRONEOUS;
+        }
+
+        boolean handled = false;
+        
+        //execute all exceptionHandlers if the type of wrapped is equal to the exceptionClass the handler handles.
+        if (p.onException != null) {
+          for (int i = 0; i < p.onException.size(); i++) {
+            SClass exceptionClass = p.onException.get(i);
+            PromiseMessage handle = p.onExceptionCallbacks.get(i);
+            if (((SAbstractObject) wrapped).getSOMClass() == exceptionClass) {
+                handled = true;
+                p.scheduleCallbacksOnResolution(exception, handle, current, false);
+            }
+          }
+        }
+
+        //execute all errorHandlers
+        if (p.onError != null) {
+          for (int i = 0; i < p.onError.size(); i++) {
+            handled = true;
+            PromiseMessage handle = p.onError.get(i);
+            p.scheduleCallbacksOnResolution(exception, handle, current, false);
+          }
+        }
+        return handled;
+    }
+
+    protected static void ruinChainedPromisesUnsynch(final SAbstractObject exception, final SPromise promise, final Actor current){
```

Comment:
Would it be possible to use the existing code here instead, and pass as an argument whether the resolution is successful or erroneous? This seems to be a duplication of rather intricate code, which I'd really like to avoid.